sys/net/gnrc/network_layer/ipv6/nib/nib.c

```diff
@@ -175,19 +179,71 @@ int gnrc_ipv6_nib_get_next_hop_l2addr(const ipv6_addr_t *dst,
     gnrc_netif2_acquire(netif);
     mutex_lock(&_nib_mutex);
     do {    /* XXX: hidden goto ;-) */
-        if (ipv6_addr_is_link_local(dst)) {
-            /* TODO: Prefix-based on-link determination */
+        _nib_onl_entry_t *node = _nib_onl_get(dst,
+                                              (netif == NULL) ? 0 : netif->pid);
+        /* consider neighbor cache entries first */
+        unsigned iface = (node == NULL) ? 0 : _nib_onl_get_if(node);
+
+        if ((node != NULL) || _on_link(dst, &iface)) {
+            DEBUG("nib: %s is on-link or in NC, start address resolution\n",
+                  ipv6_addr_to_str(addr_str, dst, sizeof(addr_str)));
+            /* on-link prefixes return their interface */
+            if (!ipv6_addr_is_link_local(dst) && (iface != 0)) {
+                /* release preassumed interface */
+                gnrc_netif2_release(netif);
+                netif = gnrc_netif2_get_by_pid(iface);
+                gnrc_netif2_acquire(netif);
+            }
             if ((netif == NULL) ||
-                !_resolve_addr(dst, netif, pkt, nce,
-                               _nib_onl_get(dst, netif->pid))) {
+                !_resolve_addr(dst, netif, pkt, nce, node)) {
+                DEBUG("nib: host unreachable\n");
                 res = -EHOSTUNREACH;
                 break;
             }
         }
         else {
-            /* TODO: Off-link next hop determination;
-             *       might need netif locking */
-            res = -EHOSTUNREACH;
+            gnrc_ipv6_nib_ft_t route;
+
+            DEBUG("nib: %s is off-link, resolve route\n",
+                  ipv6_addr_to_str(addr_str, dst, sizeof(addr_str)));
+            res = _nib_get_route(dst, pkt, &route);
+            if ((res < 0) || ipv6_addr_is_unspecified(&route.next_hop)) {
+                DEBUG("nib: no route to %s found or is prefix list entry, "
+                      "search neighbor cache\n",
+                      ipv6_addr_to_str(addr_str, dst, sizeof(addr_str)));
+
+                if (res == 0) {
+                    DEBUG("nib: prefix list entry => taking dst as next hop\n");
+                    memcpy(&route.next_hop, dst, sizeof(route.next_hop));
+                }
+                else {
+                    res = -ENETUNREACH;
+                    break;
+                }
+            }
+            if ((netif != NULL) && (netif->pid != route.iface)) {
+                /* drop pre-assumed netif */
+                gnrc_netif2_release(netif);
+            }
+            if ((netif == NULL) || (netif->pid != route.iface)) {
+                /* get actual netif */
+                netif = gnrc_netif2_get_by_pid(route.iface);
+                gnrc_netif2_acquire(netif);
```

Comment:
Is it possible to get `NULL` as a netif here? (I think not because the route is checked above, but just to be certain)

Reply:
Even if: `gnrc_netif2_acquire(netif)` checks that.

Reply:
Okay, my bad then.